go/identify3/adapter.go

```diff
@@ -0,0 +1,340 @@
+		ui:               ui,
+	}
+	return ret, nil
+}
+
+func NewUIAdapterWithSession(mctx libkb.MetaContext, ui keybase1.Identify3UiInterface, sess *libkb.Identify3Session) (*UIAdapter, error) {
+	ret, err := NewUIAdapter(mctx, ui)
+	if err != nil {
+		return nil, err
+	}
+	ret.session = sess
+	return ret, nil
+}
+
+func (i *UIAdapter) Start(string, keybase1.IdentifyReason, bool) error {
+	return nil
+}
+
+// return true if we need an upgrade
+func (i *UIAdapter) setRowStatus(arg *keybase1.Identify3UpdateRowArg, lcr keybase1.LinkCheckResult) bool {
+
+	needUpgrade := false
+	i.M().CDebugf("setRowStatus(%+v)", lcr)
+
+	switch {
+
+	// The proof worked, and either we tracked it as working, or we tracked it as working
+	case (lcr.ProofResult.State == keybase1.ProofState_OK && (lcr.RemoteDiff == nil || lcr.RemoteDiff.Type == keybase1.TrackDiffType_NONE)):
+		arg.Color = keybase1.Identify3RowColor_GREEN
+		arg.State = keybase1.Identify3RowState_VALID
+
+	// The proof worked, and it's new to us.
+	case lcr.ProofResult.State == keybase1.ProofState_OK && lcr.RemoteDiff != nil && lcr.RemoteDiff.Type == keybase1.TrackDiffType_NEW:
+		arg.Color = keybase1.Identify3RowColor_BLUE
+		arg.State = keybase1.Identify3RowState_VALID
+		arg.Metas = append(arg.Metas, keybase1.Identify3RowMeta("new"))
+		needUpgrade = true
+
+	// The proof worked, and it's upgraded.
+	case lcr.ProofResult.State == keybase1.ProofState_OK && lcr.RemoteDiff != nil && lcr.RemoteDiff.Type == keybase1.TrackDiffType_UPGRADED:
+		arg.Color = keybase1.Identify3RowColor_BLUE
+		arg.State = keybase1.Identify3RowState_VALID
+		arg.Metas = append(arg.Metas, keybase1.Identify3RowMeta("upgraded"))
+		needUpgrade = true
+
+	// The proof worked, we tracked failed, and now it's working
+	case lcr.ProofResult.State == keybase1.ProofState_OK && lcr.RemoteDiff != nil && lcr.RemoteDiff.Type == keybase1.TrackDiffType_REMOTE_WORKING:
+		arg.Color = keybase1.Identify3RowColor_BLUE
+		arg.State = keybase1.Identify3RowState_VALID
+		arg.Metas = append(arg.Metas, keybase1.Identify3RowMeta("upgraded"))
+		needUpgrade = true
+
+	// The proof failed, but we didn't track it
+	case lcr.ProofResult.State != keybase1.ProofState_OK && lcr.RemoteDiff == nil:
+		arg.Color = keybase1.Identify3RowColor_ORANGE
+		arg.State = keybase1.Identify3RowState_WARNING
+		arg.Metas = append(arg.Metas, keybase1.Identify3RowMeta("unreachable"))
+
+	// The proof failed, but we did "ignore" it, so it's OK
+	case lcr.ProofResult.State != keybase1.ProofState_OK && (lcr.RemoteDiff == nil || lcr.RemoteDiff.Type == keybase1.TrackDiffType_NONE):
+		arg.Color = keybase1.Identify3RowColor_ORANGE
+		arg.State = keybase1.Identify3RowState_WARNING
+		arg.Metas = append(arg.Metas, keybase1.Identify3RowMeta("ignored"))
+
+	// The proof failed, but we did "ignore" it via "SNOONZE", so it's OK, for now.
+	case lcr.ProofResult.State != keybase1.ProofState_OK && lcr.RemoteDiff != nil && lcr.RemoteDiff.Type == keybase1.TrackDiffType_NONE_VIA_TEMPORARY:
```

Comment:
again, i think you know that lcr.RemoteDiff isn't nil already.

Reply:
sure, i was just trying to read through all the cases and see what was different.

Reply:
thanks for the careful review!